maps/src/main/java/org/odk/collect/maps/layers/OfflineMapLayersImporterViewModel.kt

```diff
@@ -0,0 +1,65 @@
+package org.odk.collect.maps.layers
+
+import android.content.ContentResolver
+import android.net.Uri
+import androidx.lifecycle.LiveData
+import androidx.lifecycle.MutableLiveData
+import androidx.lifecycle.ViewModel
+import org.odk.collect.androidshared.system.getFileName
+import org.odk.collect.androidshared.system.toFile
+import org.odk.collect.async.Scheduler
+import org.odk.collect.shared.TempFiles
+import java.io.File
+import java.util.ArrayList
+
+class OfflineMapLayersImporterViewModel(
+    private val scheduler: Scheduler,
+    private val contentResolver: ContentResolver
+) : ViewModel() {
+
+    private val _data = MutableLiveData<List<ReferenceLayer>>()
+    val data: LiveData<List<ReferenceLayer>> = _data
+
+    private lateinit var tempLayersDir: File
+
+    fun init(uris: ArrayList<String>?) {
+        scheduler.immediate(
+            background = {
+                tempLayersDir = TempFiles.createTempDir().also {
+                    it.deleteOnExit()
+                }
+                val layers = mutableListOf<ReferenceLayer>()
+                uris?.forEach { uriString ->
+                    val uri = Uri.parse(uriString)
+                    uri.getFileName(contentResolver)?.let { fileName ->
+                        if (fileName.endsWith(MbtilesFile.FILE_EXTENSION)) {
+                            val layerFile = File(tempLayersDir, fileName).also { file ->
+                                uri.toFile(contentResolver, file)
+                            }
+                            layers.add(ReferenceLayer(layerFile.absolutePath, layerFile, MbtilesFile.readName(layerFile) ?: layerFile.name))
+                        }
+                    }
+                }
+                _data.postValue(layers)
+            },
+            foreground = { }
+        )
+    }
+
+    fun addLayers(layersDir: String): LiveData<Boolean> {
```

Comment:
Ideally the directory path itself could be hidden from the UI and this could just take boolean (to flag shared or project).

I'm realizing that it might make sense to just have one ViewModel shared between the two Fragments here. That would work with other comments (like moving where `init` is called), and would remove the need to use Fragment results to signal to `OfflineMapLayersPicker` to refresh - the ViewModel would just refresh its list of layers when one is added.

Reply:
One instead of the existing two (`OfflineMapLayersPickerViewModel` and `OfflineMapLayersImporterViewModel`) or a third one?

Reply:
Just one. I'm thinking something like `OfflineMapLayersViewModel` with an interface like:

```kotlin
val layers: LiveData<Pair<List<ReferenceLayer>, String?>?>
val isLoading: LiveData<Boolean>
val filesToImport: LiveData<List<ReferenceLayer>?>
fun changeSelectedLayerId(selectedLayerId: String?)
fun saveSelectedLayer()
fun importLayers(projectOnly: Boolean)
fun addExternalFiles(uris: ArrayList<String>?)
```

Alternatively, we could keep two, and just have second expose `filesToImport` and `addExternalFiles` and have the shared one take a list of reference layers to import in `importLayers`.

Reply:
Done.